File: backend/app/services/moodle_service.py
"""
Moodle integration service
"""
import httpx
from typing import Dict, Any, Optional, List
import logging

from app.core.config import get_settings

logger = logging.getLogger(__name__)
settings = get_settings()


class MoodleService:
    """Service for Moodle API integration"""

    def __init__(self):
        self.base_url = settings.MOODLE_URL.rstrip('/')
        self.token = settings.MOODLE_TOKEN
        self.ws_url = f"{self.base_url}/webservice/rest/server.php"

    async def call_function(self, function: str, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Call a Moodle web service function

        Args:
            function: Moodle web service function name
            params: Function parameters

        Returns:
            Response data
        """
        if not self.token:
            logger.warning("Moodle token not configured")
            return {}

        request_params = {
            'wstoken': self.token,
            'wsfunction': function,
            'moodlewsrestformat': 'json'
        }

        if params:
            request_params.update(params)

        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(self.ws_url, data=request_params)
                response.raise_for_status()
                return response.json()

        except httpx.HTTPError as e:
            logger.error(f"Moodle API error: {str(e)}")
            raise
        except Exception as e:
            logger.error(f"Error calling Moodle function {function}: {str(e)}")
            raise

    async def get_user_info(self, user_id: int) -> Optional[Dict[str, Any]]:
        """Get user information from Moodle"""
        try:
            result = await self.call_function(
                'core_user_get_users_by_field',
                {'field': 'id', 'values[0]': user_id}
            )
            return result[0] if result else None
        except Exception as e:
            logger.error(f"Error getting user info: {str(e)}")
            return None

    async def get_assignment_submissions(self, assignment_id: int) -> List[Dict[str, Any]]:
        """Get assignment submissions from Moodle"""
        try:
            result = await self.call_function(
                'mod_assign_get_submissions',
                {'assignmentids[0]': assignment_id}
            )
            return result.get('assignments', [{}])[0].get('submissions', [])
        except Exception as e:
            logger.error(f"Error getting submissions: {str(e)}")
            return []

    async def get_course_students(self, course_id: int) -> List[Dict[str, Any]]:
        """Get students enrolled in a course"""
        try:
            result = await self.call_function(
                'core_enrol_get_enrolled_users',
                {'courseid': course_id}
            )
            return result if isinstance(result, list) else []
        except Exception as e:
            logger.error(f"Error getting course students: {str(e)}")
            return []

    async def update_grade(self, assignment_id: int, user_id: int, grade: float, feedback: str = "") -> bool:
        """Update assignment grade in Moodle"""
        try:
            await self.call_function(
                'mod_assign_save_grade',
                {
                    'assignmentid': assignment_id,
                    'userid': user_id,
                    'grade': grade,
                    'plugindata[assignfeedbackcomments_editor][text]': feedback,
                    'plugindata[assignfeedbackcomments_editor][format]': 1
                }
            )
            return True
        except Exception as e:
            logger.error(f"Error updating grade: {str(e)}")
            return False

    async def send_feedback(self, user_id: int, subject: str, message: str) -> bool:
        """Send message to user via Moodle"""
        try:
            await self.call_function(
                'core_message_send_instant_messages',
                {
                    'messages[0][touserid]': user_id,
                    'messages[0][text]': message,
                    'messages[0][textformat]': 1
                }
            )
            return True
        except Exception as e:
            logger.error(f"Error sending feedback: {str(e)}")
            return False

File: backend/services/MoodleService.php
<?php
/**
 * Moodle LMS Integration Service
 * Moodle 3.7 Web Services API
 */

require_once __DIR__ . '/../config/config.php';

class MoodleService {
    private $moodleUrl;
    private $token;
    private $restFormat = 'json';

    public function __construct() {
        $this->moodleUrl = MOODLE_URL;
        $this->token = MOODLE_TOKEN;
    }

    /**
     * Call Moodle Web Service
     */
    private function callMoodleAPI($function, $params = []) {
        $serverUrl = $this->moodleUrl . '/webservice/rest/server.php';

        $requestParams = array_merge([
            'wstoken' => $this->token,
            'wsfunction' => $function,
            'moodlewsrestformat' => $this->restFormat
        ], $params);

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $serverUrl);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($requestParams));
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false); // For development only

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($error) {
            $this->logError("Moodle API cURL error: " . $error);
            return ['error' => $error];
        }

        if ($httpCode !== 200) {
            $this->logError("Moodle API HTTP error: " . $httpCode);
            return ['error' => 'HTTP ' . $httpCode];
        }

        $data = json_decode($response, true);

        if (isset($data['exception'])) {
            $this->logError("Moodle API exception: " . $data['message']);
            return ['error' => $data['message']];
        }

        return $data;
    }

    /**
     * Get user information
     */
    public function getUser($userId) {
        return $this->callMoodleAPI('core_user_get_users_by_field', [
            'field' => 'id',
            'values[0]' => $userId
        ]);
    }

    /**
     * Get course information
     */
    public function getCourse($courseId) {
        return $this->callMoodleAPI('core_course_get_courses', [
            'options[ids][0]' => $courseId
        ]);
    }

    /**
     * Get quiz questions
     */
    public function getQuizQuestions($quizId) {
        return $this->callMoodleAPI('mod_quiz_get_quiz_access_information', [
            'quizid' => $quizId
        ]);
    }

    /**
     * Get enrolled users in a course
     */
    public function getEnrolledUsers($courseId) {
        return $this->callMoodleAPI('core_enrol_get_enrolled_users', [
            'courseid' => $courseId
        ]);
    }

    /**
     * Submit quiz attempt (for progress tracking)
     */
    public function submitQuizAttempt($attemptId, $data) {
        return $this->callMoodleAPI('mod_quiz_save_attempt', [
            'attemptid' => $attemptId,
            'data' => json_encode($data)
        ]);
    }

    /**
     * Sync students from Moodle
     */
    public function syncStudents($courseId) {
        $db = Database::getInstance();
        $users = $this->getEnrolledUsers($courseId);

        if (isset($users['error'])) {
            return ['success' => false, 'error' => $users['error']];
        }

        $syncedCount = 0;
        $db->beginTransaction();

        try {
            foreach ($users as $user) {
                $sql = "INSERT INTO students (moodle_user_id, username, email, full_name)
                        VALUES (:moodle_user_id, :username, :email, :full_name)
                        ON DUPLICATE KEY UPDATE
                        username = :username, email = :email, full_name = :full_name, updated_at = NOW()";

                $params = [
                    ':moodle_user_id' => $user['id'],
                    ':username' => $user['username'],
                    ':email' => $user['email'] ?? '',
                    ':full_name' => $user['fullname'] ?? ''
                ];

                $db->query($sql, $params);
                $syncedCount++;
            }

            $db->commit();
            $this->logSync('students', 'success', $syncedCount);

            return ['success' => true, 'synced' => $syncedCount];

        } catch (Exception $e) {
            $db->rollback();
            $this->logSync('students', 'failed', 0, $e->getMessage());
            return ['success' => false, 'error' => $e->getMessage()];
        }
    }

    /**
     * Sync problems/questions from Moodle
     */
    public function syncProblems($quizId, $courseId) {
        $db = Database::getInstance();
        $quiz = $this->getQuizQuestions($quizId);

        if (isset($quiz['error'])) {
            return ['success' => false, 'error' => $quiz['error']];
        }

        // This is a simplified version - actual implementation would need
        // to use mod_quiz_get_attempt_data or similar to get question details

        $syncedCount = 0;
        $this->logSync('problems', 'success', $syncedCount);

        return ['success' => true, 'synced' => $syncedCount];
    }

    /**
     * Log synchronization
     */
    private function logSync($type, $status, $count, $error = null) {
        $db = Database::getInstance();
        $sql = "INSERT INTO moodle_sync_log (sync_type, status, records_synced, error_message, sync_completed_at)
                VALUES (:type, :status, :count, :error, NOW())";

        $db->query($sql, [
            ':type' => $type,
            ':status' => $status,
            ':count' => $count,
            ':error' => $error
        ]);
    }

    /**
     * Log errors
     */
    private function logError($message) {
        error_log("[MoodleService] " . $message);
    }
}
